[JIT] NamedTuple serialization

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21839 [JIT] NamedTuple serialization**
* #21836 [JIT] Refactor TupleType to take a FunctionSchema
* #21835 [JIT] Introduce NamedType
* #21813 [JIT] Some small fixes for NamedTuple

Differential Revision: [D15860002](https://our.internmc.facebook.com/intern/diff/D15860002)

Commit:
[JIT] NamedTuple serialization

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -845,36 +845,88 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
+struct SerializableType;
+using SerializableTypePtr = std::shared_ptr<SerializableType>;
+
+struct CAFFE2_API SerializableType : public Type {
```
Huh, wasn't this in the other PR?

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -845,36 +845,88 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
+    return name_.has_value();
+  }
+
+  virtual c10::optional<std::string> base_class_name() const = 0;
```
This API feels pretty overspecialized to the two classes that are supported here. Interface, for instance, would not be able to work in the API because they do not come with Functions,  just FunctionSchema.  What is stopping us from just having a switch statement in the serializer to handle NamedTuple and ClassType differently?

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -52,8 +53,12 @@ struct TORCH_API ClassNamespaceValue : public SugaredValue {
     auto fullName = c10::QualifiedName(basename_, name);
-    if (auto classType = cu_.get_class(fullName)) {
-      return std::make_shared<ClassValue>(classType);
+    if (auto serializable_type = cu_.get_serializable_type(fullName)) {
```
This is going to conflict with https://github.com/pytorch/pytorch/pull/21853 which also needs to redefine how named types are put in the compilation unit.

zdevito replied:
`get_serializabe_type` -> `get_type`

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -141,7 +146,7 @@ struct SourceResolver : public Resolver {
 
   TypePtr resolveType(const std::string& name, const SourceRange& loc) const override {
-    return lib_cu_.get_class(c10::QualifiedName(name));
+    return lib_cu_.get_serializable_type(c10::QualifiedName(name));
```
Yeah, we should coordinate how this will work across interfaces and named tuples, both require extending this API.

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -184,23 +189,56 @@ struct SourceImporter {
+            TupleType::NamedTupleSpec(field_types, field_names));
+        owner.register_class(tt);
+      } else {
+        TORCH_INTERNAL_ASSERT(false);
```
maybe a more descriptive error?

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -184,23 +189,56 @@ struct SourceImporter {
-        definitions.emplace_back(method_def);
-        resolvers.emplace_back(resolver_);
-      }
+      auto parsed_treeref = p_.parseClass();
```
parseClassLike? make it more clear it might return a named tuple or an interface.

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1104,10 +1114,20 @@ struct PythonPrintPass {
 
   void printClass(const c10::SerializableTypePtr& classType) {
-    body_ << "class " << classType->basename() << ":\n";
+    body_ << "class " << classType->basename();
```
I think it would be fine here to simply have another switch and then use helper functions to emit the body. The problem with the current way is that it won't support interfaces anyway, so we'd need to introduce the switch statement still, and the parser part already does it with a switch. So it is a bit confusing that input is doing via switches but output is done via virtual methods. I like the idea of making NamedTuple/Class/Interface all inherit from NamedTypePtr, which could then be what goes in the compilation units type table.

zdevito commented on torch/csrc/jit/script/lexer.h:
```diff
@@ -100,7 +100,9 @@ namespace script {
   _(TK_CLASS_DEF, "class", "class")              \
-  _(TK_IMPORT, "import", "import")
+  _(TK_IMPORT, "import", "import")               \
+  _(TK_NAMED_TUPLE, "NamedTuple", "NamedTuple")  \
```
This should not be added as a token because it is not a python keyword.  It prevents someone from writing:

```
def foo():
   NamedTuple = 3
   return NamedTuple
```

which, while weird, is perfectly valid code.

zdevito commented on torch/csrc/jit/script/parser.cpp:
```diff
@@ -573,14 +581,37 @@ struct ParserImpl {
-      L.reportError(
-          "Inheritance is not yet supported for TorchScript classes yet.");
+      // Only support inheriting from NamedTuple right now.
+      if (L.nextIf(TK_NAMED_TUPLE)) {
```
`L.nextIf(TK_IDENT).value() == "NamedTuple"`

zdevito commented on torch/csrc/jit/script/parser.cpp:
```diff
@@ -573,14 +581,37 @@ struct ParserImpl {
+        return parseNamedTuple(name);
+      } else {
+        L.reportError(
+            "Inheritance is not yet supported for TorchScript classes yet.");
```
remove the yet, don't give people hope!

zdevito reviewed (commented):
This looks pretty good. Let's coordinate a few things tomorrow morning to make sure it works for both Interfaces and NamedTuples

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

zdevito reviewed (approved):
Looks good!

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head

Commit:
Update on "[JIT] NamedTuple serialization"

[JIT] NamedTuple serialization

gh-metadata: pytorch pytorch 21839 gh/jamesr66a/6/head
